database/service_instance.go

```diff
@@ -0,0 +1,563 @@
+	getInput := &GetServiceInstanceInput{
+		Name: name,
+	}
+
+	// Wait for the service instance to be running and return the result
+	// Don't have to unqualify any objects, as the GetServiceInstance method will handle that
+	serviceInstance, serviceInstanceError := c.WaitForServiceInstanceRunning(getInput, WaitForServiceInstanceReadyTimeout)
+	// If the service instance enters an error state we need to delete the instance and retry
+	if serviceInstanceError != nil {
+		deleteInput := &DeleteServiceInstanceInput{
+			Name: name,
+		}
+		err := c.DeleteServiceInstance(deleteInput)
+		if err != nil {
+			return nil, fmt.Errorf("Error deleting service instance %s: %s", name, err)
+		}
+		return nil, serviceInstanceError
+	}
+	return serviceInstance, nil
+}
+
+// WaitForServiceInstanceRunning waits for a service instance to be completely initialized and available.
+func (c *ServiceInstanceClient) WaitForServiceInstanceRunning(input *GetServiceInstanceInput, timeoutSeconds int) (*ServiceInstance, error) {
+	var info *ServiceInstance
+	var getErr error
+	err := c.client.WaitFor("service instance to be ready", timeoutSeconds, func() (bool, error) {
+		info, getErr = c.GetServiceInstance(input)
+		if getErr != nil {
+			return false, getErr
+		}
+		c.client.DebugLogString(fmt.Sprintf("Service instance name is %v, Service instance info is %+v", info.Name, info))
+		switch s := info.Status; s {
+		case ServiceInstanceRunning: // Target State
+			c.client.DebugLogString("Service Instance Running")
+			return false, nil
+		case ServiceInstanceConfigured:
+			c.client.DebugLogString("Service Instance Configured")
+			return true, nil
+		case ServiceInstanceInProgress:
+			c.client.DebugLogString("Service Instance is being created")
+			return false, nil
+		default:
+			c.client.DebugLogString(fmt.Sprintf("Unknown instance state: %s, waiting", s))
+			return false, nil
+		}
+	})
+	return info, err
+}
+
+type GetServiceInstanceInput struct {
+	// Name of the Database Cloud Service instance.
+	// Required.
+	Name string `json:"serviceId"`
+}
+
+// GetServiceInstance retrieves the SeriveInstance with the given name.
+func (c *ServiceInstanceClient) GetServiceInstance(getInput *GetServiceInstanceInput) (*ServiceInstance, error) {
+	var serviceInstance ServiceInstance
+	if err := c.getResource(getInput.Name, &serviceInstance); err != nil {
+		return nil, err
+	}
+
+	return &serviceInstance, nil
+}
+
+type DeleteServiceInstanceInput struct {
+	// Name of the Database Cloud Service instance.
+	// Required.
+	Name string `json:"serviceId"`
+}
+
+func (c *ServiceInstanceClient) DeleteServiceInstance(deleteInput *DeleteServiceInstanceInput) error {
+	// Call to delete the service instance
+	// If delete fails, rerun it in case the instance still has not been setup correctly.
+	// An instance takes additional time to setup after it's configured.
+	var deleteErr error
+	for i := 0; i < 5; i++ {
```

Comment:
This should be set to the client's max_retries parameter, not an arbitrary integer :smile: